### PR TITLE
core_response: new jResponseStreamed

### PR DESCRIPTION
**Pull request by rldhont**
This new response jResponseStreamed is inspired by Symphony `StreamedResponse` https://symfony.com/doc/current/components/http_foundation.html#streaming-a-response

This new response has been created to streamed content without to build a file before sending the content. For exemple to send a CSV content build from an SQL request, if the number of
lines is too heavy, it is necessary to write a file before sending the content and then dropping file.

This new response has been imagined to send a part of a file based on HTTPRange header or to send a file by chunk.

To use this new response.

* You can set the callback

```php
$rep = $this->getResponse('streamed');
$rep->mimeType = 'text/plain';
$rep->setCallback(function () {
    $iterator = range(0, 12);
    foreach ($iterator as $line) {
        sleep(1);
        echo $line;
    }
 });
return $rep;
```

* You can set the callback with use external object

```php
$rep = $this->getResponse('streamed');
$rep->mimeType = 'text/plain';
$iterator = range(0, 12);
$rep->setCallback(function () use ($iterator) {
    foreach ($iterator as $line) {
        sleep(1);
        echo $line;
    }
 });
return $rep;
```

* You can use an iterable

```php
$rep = $this->getResponse('streamed');
$rep->mimeType = 'text/plain';
$iterator = range(0, 12);
$rep->setGenerator($iterator);
return $rep;
```

* You can use a generator to get an iterable

```php
function gen($iterator) {
    foreach ($iterator as $line) {
        sleep(1);
        yield $line;
    }
}
$rep = $this->getResponse('streamed');
$rep->mimeType = 'text/plain';
$rep->setGenerator(gen($iterator));
return $rep;
```

* You can use an anonymous generator

```php
$rep = $this->getResponse('streamed');
$rep->mimeType = 'text/plain';
$rep->setGenerator((function () use ($iterator){
    foreach ($iterator as $line) {
        sleep(1);
        yield $line;
    }
})());
return $rep;
```

* You can use a class generator method

```php
protected function xrange($start, $limit, $step = 1) {
    if ($start <= $limit) {
        if ($step <= 0) {
            throw new LogicException('Step must be positive');
        }

        for ($i = $start; $i <= $limit; $i += $step) {
            yield $i;
        }
    } else {
        if ($step >= 0) {
            throw new LogicException('Step must be negative');
        }

        for ($i = $start; $i >= $limit; $i += $step) {
            yield $i;
        }
    }
}

public function getrange()
{
    $rep = $this->getResponse('streamed');
    $rep->mimeType = 'text/plain';

    $start = $this->param('start');
    $end = $this->param('end');
    $step = $this->param('step', 1);

    $rep->setGenerator($this->xrange($start, $end, $step));
    return $rep;
}
```

### CODE REVIEW

File: lib/jelix/core/response/jResponseStreamed.class.php [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+    /**
+     * @var callable The callback use to send the content
+     */
+    protected $callback;
+
+    /**
+     * @var bool The content is streamed
+     */
+    protected $streamed;
```
Comment (laurentj) [inline lib/jelix/core/response/jResponseStreamed.class.php:53]: it seems it is useless?

File: lib/jelix/core/response/jResponseStreamed.class.php [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+    /**
+     * @var bool The header has been sent
+     */
+    private $headersSent;
```
Comment (laurentj) [inline lib/jelix/core/response/jResponseStreamed.class.php:58]: There is already a `_httpHeadersSent` property that exists on the jResponse class, and modified by the `sendHttpHeaders()` method.  You should remove it.

File: lib/jelix/core/response/jResponseStreamed.class.php [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+            $this->sendHttpHeaders();
+
+            return true;
+        }
+        $this->sendHttpHeaders();
+        $this->headersSent = true;
+
+        $this->streamed = true;
+
+        if (null === $this->callback) {
```
Comment (laurentj) [inline lib/jelix/core/response/jResponseStreamed.class.php:81]: this test should be done before calling `$this->sendHttpHeaders();`, so headers for the error can be set correctly by the core of Jelix.

File: lib/jelix/core/response/jResponseStreamed.class.php [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+        return true;
+    }
+
+    /**
+     * Sets the PHP callback associated with this Response.
+     *
+     * @param callable $callback The callback use to send the content
+     *
+     */
+    public function setCallback($callback)
```
Comment (laurentj) [inline lib/jelix/core/response/jResponseStreamed.class.php:96]: force the parameter type, or check the type yourself into the method: `public function setCallback(callable $callback)`

File: lib/jelix/core/response/jResponseStreamed.class.php [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+    }
+
+    /**
+     * Sets the PHP callback associated with this Response with an
+     * iterator
+     *
+     * @param iterable $iterator The result of a generator use to build the callback to send the content
+     *
+     */
+    public function setGenerator($iterator)
```
Comment (laurentj) [inline lib/jelix/core/response/jResponseStreamed.class.php:108]: force the type of the parameter : `public function setGenerator(iterable $iterator)`

**Review by laurentj — CHANGES_REQUESTED** [review]

**Comment by rldhont** [comment]
@laurentj thanks for the review. I have updated the class.

**Review by laurentj — APPROVED** [review]
thank you

**Comment by rldhont** [comment]
Thanks @laurentj

**Pull request merged**